memcache/selector_test.go

```diff
@@ -16,24 +16,284 @@ limitations under the License.
+		benchPickServer(t, "127.0.0.1:1234", "127.0.0.1:1235")
+	})
+	t.Run("no allocations with one server", func(t *testing.T) {
+		benchPickServer(t, "127.0.0.1:1234")
+	})
 }
 
-func BenchmarkPickServer_Single(b *testing.B) {
-	benchPickServer(b, "127.0.0.1:1234")
+
+func benchPickServer(t *testing.T, servers ...string) {
+	br := testing.Benchmark(func(b *testing.B) {
+		b.ReportAllocs()
+		var ss ServerList
+		requireNoError(b, ss.SetServers(servers...))
+		for i := 0; i < b.N; i++ {
+			if _, err := ss.PickServer("some key"); err != nil {
+				b.Fatal(err)
+			}
+		}
+	})
+	if br.AllocsPerOp() > 0 {
+		t.Errorf("expected no allocations, got %d", br.AllocsPerOp())
+	}
+}
+
+// Note: this test is technically non-deterministic as the retry scheduling is async,
+// but failures should be exceedingly rare due to the length of time until the first timeout.
+//
+// If a fix is required: probably simplest to change the startingWait to a var and modify in this test
+func TestBehaviourSynchronously(t *testing.T) {
+	create := func(t *testing.T) (ServerSelector, net.Addr, net.Addr) {
+		ss := createWithServers(t, []string{"127.0.0.1:4200", "127.255.0.1:4200"})
+		addrA := ss.addrs[0]
+		addrB := ss.addrs[1]
+		if addrA == addrB {
+			t.Fatalf("for test to be valid addresses must differ")
+		}
+		return ss, addrA, addrB
+	}
+
+	t.Run("after a net error, an address should be not be pickable until retried", func(t *testing.T) {
+		ss, addrA, _ := create(t)
+
+		ss.OnResult(addrA, &net.OpError{})
+		fuzzer(func(k string) {
+			addr, err := ss.PickServer(k)
+			requireNoError(t, err)
+			if addrA == addr {
+				t.Fatalf("address still available")
+			}
+		})
+	})
+
+	t.Run("after a successful retry, an address should be available again", func(t *testing.T) {
+		ss, addrA, _ := create(t)
+		ss.OnResult(addrA, nil)
+		gotAddr := 0
+		fuzzer(func(k string) {
+			addr, err := ss.PickServer(k)
+			requireNoError(t, err)
+			if addr == addrA {
+				gotAddr++
+			}
+		})
+		if gotAddr == 0 {
+			t.Fatalf("address not pickable again")
+		}
+	})
+
+	t.Run("after a net error, an address should be removed", func(t *testing.T) {
+		ss, addrA, addrB := create(t)
+		// when all address have errors, we return an error that signals no servers are available (circuit breaker)
+		ss.OnResult(addrA, &net.OpError{})
+		ss.OnResult(addrB, &net.OpError{})
+
+		_, err := ss.PickServer("hi")
+		assertEqualError(t, err, "memcache: no servers configured or available")
+	})
 }
 
-func benchPickServer(b *testing.B, servers ...string) {
-	b.ReportAllocs()
-	var ss ServerList
-	ss.SetServers(servers...)
-	for i := 0; i < b.N; i++ {
-		if _, err := ss.PickServer("some key"); err != nil {
-			b.Fatal(err)
+func TestRetriesAfterError(t *testing.T) {
+	ss := createWithServers(t, []string{"127.0.0.1:1234"})
+
+	addrA := ss.addrs[0]
+	ss.OnResult(addrA, &net.OpError{})
+	_, err := ss.PickServer("hi")
+	assertEqualError(t, err, "memcache: no servers configured or available")
+
```

Comment:
not sure its worth it, but we could also sleep for a short period of time and expect for `PickServer` to return an error?

[also, is it worth making any assertion about `ss` internals? Right now we are testing the interface, which seems to be a good approach, but there may also be some increased fidelity to actualy verifying the state tracking -- I imagine that would make regressions to the concurrency logic for example easier to catch later (requires us to be comfortable with the test peeking under the hood of things a bit)].

Reply:
👍 

Have written up some tests of `filterAvailable`, `setState` and `deleteState`, and caught an issue where load-balanced servers (i.e those present in `addrs` multiple times) weren't being handled properly.